cmd_watch.go

```diff
@@ -103,48 +109,99 @@ build with the appropriate timers.`,
 	return watchCmd
 }
 
+// waitCircle polls the CircleCI API checking for the status of this workflow
+// and the jobs it contains. It returns whether the workflow build succeeded,
+// the time it started. The err returned is for errors polling the CircleCI API,
+// not errors in the build itself.
 func waitCircle(parent context.Context, cfg watchConfig) (bool, time.Time, error) {
+	// we need a token to query anything; give a helpful error if we have no token
+	if cfg.circleKey == "" {
+		return false, time.Now(), fmt.Errorf("circle token required to poll the API")
+	}
 	client := &circleci.Client{Token: cfg.circleKey}
 	wf, err := client.GetWorkflowV2(cfg.workflowID)
 	if err != nil {
 		return false, time.Now(), err
 	}
 	start := wf.CreatedAt
-	var failed bool
+	var passed bool
 
 	done := make(chan struct{})
 	ctx, cxl := context.WithTimeout(parent, time.Duration(cfg.timeoutMin)*time.Minute)
 	defer cxl()
 
+	// sometimes there's a gap between when a job finishes and the next one starts.
+	// In that case there are no jobs running and some jobs blocked that could
+	// still run. If we think the build has passed and finished, let's give it a
+	// buffer to spin up new jobs before really considering it done.
+	checksLeft := numChecks + 1 // +1 because we decrement at the beginning of the loop
+
 	go func() {
 		defer close(done)
-		for {
+		tk := time.NewTicker(5 * time.Second).C
+		for range tk {
+			// check for timeout or pause before the next iteration
 			select {
 			case <-ctx.Done():
-				failed = true
+				// TODO add the fact that it timed out to the trace to say why it failed
+				fmt.Fprintf(os.Stderr, "Timeout reached waiting for the workflow to finish\n")
 				return
-			case <-time.After(5 * time.Second):
-				// move on thru
+			default:
 			}
 
-			var finished bool
-			finished, failed = evalWorkflow(client, cfg.workflowID, cfg.jobName)
+			finished, failed, err := evalWorkflow(client, cfg.workflowID, cfg.jobName)
 			if finished {
-				return
+				checksLeft--
+				if checksLeft <= 0 {
```

Comment:
can't you avoid the jiggering around with adding one and then subtracting one by checking for if < 0 rather than <= 0?

Reply:
I'm using the same counter for writing to the user-visible output and counting down, and perhaps stupidly, I want the code counter to be zero-indexed and the human counter to be 1-indexed. 
```
3 more times...
2 more times...
1 more times...
done
```

rather than counting down to `0 more times...` then `done`.  I suppose I could make the code 1-indexed (aka the test would be `<1`?) but ... 

yeah, I'll give that a try.

Reply:
yeah, seems like removing the `+ 1` means we can remove the `=`

Reply:
ugh. math sucks. not gonna do this, sorry.